x-pack/apm-server/aggregation/txmetrics/aggregator.go

```diff
@@ -0,0 +1,339 @@
+	// publish a single-value metric document.
+	//
+	// TODO(axw) log a warning with a rate-limit, increment a counter.
+	counts := []int64{1}
+	values := []float64{float64(durationMicros(duration))}
+	metricset := makeMetricset(key, time.Now(), counts, values)
+	return &metricset
+}
+
+func (a *Aggregator) updateTransactionMetrics(key transactionAggregationKey, duration time.Duration) bool {
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+
+	m := a.active
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	metrics, ok := m.groups[key]
+	if !ok {
+		if len(m.groups) == a.config.MaxTransactionGroups {
+			return false
+		}
+		metrics = &transactionMetrics{
+			histogram: hdrhistogram.New(
+				durationMicros(minDuration),
+				durationMicros(maxDuration),
+				a.config.HDRHistogramSignificantFigures,
+			),
+		}
+		m.groups[key] = metrics
+	}
+	if duration < minDuration {
+		duration = minDuration
+	} else if duration > maxDuration {
+		duration = maxDuration
+	}
+	metrics.histogram.RecordValue(durationMicros(duration))
+	return true
+}
+
+// makeMetricset makes a Metricset from key, counts, and values, with timestamp ts.
+func makeMetricset(key transactionAggregationKey, ts time.Time, counts []int64, values []float64) model.Metricset {
+	out := model.Metricset{
+		Timestamp: ts,
+		Metadata: model.Metadata{
+			Service: model.Service{
+				Name:        key.serviceName,
+				Version:     key.serviceVersion,
+				Environment: key.serviceEnvironment,
+				Agent:       model.Agent{Name: key.agentName},
+			},
+			System: model.System{
+				DetectedHostname: key.hostname,
+				Container:        model.Container{ID: key.containerID},
+				Kubernetes:       model.Kubernetes{PodName: key.kubernetesPodName},
+			},
+			User: model.User{UserAgent: key.userAgent},
+			// TODO(axw) include client.geo.country_iso_code somewhere
+		},
+		Transaction: model.MetricsetTransaction{
+			Name:   key.transactionName,
+			Type:   key.transactionType,
+			Result: key.transactionResult,
+			Root:   key.traceRoot,
+		},
+		Samples: []model.Sample{{
+			Name:   "transaction.duration.histogram",
+			Counts: counts,
+			Values: values,
+		}},
+	}
+	return out
+}
+
+type metrics struct {
+	// TODO(axw) investigate optimising the map for concurrent
+	// updates, along the same lines as what we do in th Go agent
+	// for breakdown metrics.
+	mu     sync.Mutex
+	groups map[transactionAggregationKey]*transactionMetrics
+}
+
+func newMetrics() *metrics {
+	return &metrics{
+		groups: make(map[transactionAggregationKey]*transactionMetrics),
+	}
+}
+
+type transactionAggregationKey struct {
+	traceRoot bool
+	agentName string
+	// TODO(axw) requires geoIP lookup in apm-server.
+	//clientCountryISOCode string
+	containerID        string
+	hostname           string
+	kubernetesPodName  string
+	serviceEnvironment string
+	serviceName        string
+	serviceVersion     string
+	transactionName    string
+	transactionResult  string
+	transactionType    string
+	userAgent          string
+}
```

Comment:
Premature to support it now but is there a path to add more+custom dimensions to the aggregation key?

Reply:
Yes, and this will be made easier with my followup to use a custom hash map structure. We would just use a custom equality method, rather than using the compiler-generated struct equality.

Reply:
BTW @roncohen and I briefly discussed making this a libbeat processor, so it could be used by other beats.

The rough idea would be:
 - make the grouping field names, and metric field, configurable
 - output libbeat events with the grouping fields + aggregated metric field

I don't intend to do that until there's a need for it though.